File: test-verification-layer.py
#!/usr/bin/env python3
"""
Verification Layer Test Suite
Tests hash generation, response validation, and cache operations
"""

import hashlib
from typing import List

# Test Results
results = {"passed": 0, "failed": 0, "tests": []}

# Helper: Add a test case
def test(description: str, test_fn):
    try:
        result = test_fn()
        if result:
            results["passed"] += 1
            print(f"✅ {description}")
            results["tests"].append({"description": description, "status": "PASSED"})
        else:
            results["failed"] += 1
            print(f"❌ {description}")
            results["tests"].append({"description": description, "status": "FAILED"})
    except Exception as e:
        results["failed"] += 1
        print(f"❌ {description}: {e}")
        results["tests"].append({"description": description, "status": "FAILED", "error": str(e)})

# Helper: SHA-256 hash generation
def hash_content(content: str) -> str:
    return hashlib.sha256(content.encode()).hexdigest()

# -------------------------------------------------------------------------
# 1. Hash Generation Test
def test_hash_generation():
    """Test that hash is deterministic"""
    content = "This is a test string"
    hash1 = hash_content(content)
    hash2 = hash_content(content)
    return hash1 == hash2

# ---------------------------------------------------------------------
# 2. Hash Validation Test
def test_hash_validation():
    """Test hash validation works correctly"""
    # Generate hash for test content
    original_content = "This is valid training data for hash verification"
    original_hash = hash_content(original_content)
    
    # Create a modified version with one character changed
    modified_content = original_content[:5] + "X" + original_content[6:]
    modified_hash = hash_content(modified_content)
    
    # Verify: original hash should match original hash, modified should differ
    return hash1 == original_hash and hash1 != modified_hash

# -----------------------------------------------------------------------------
# 2. Response Validation Test
def test_hash_validation():
    """Simulate validation using a mock response"""
    # Mock a valid response
    response = {
        "content": "This is a valid response",
        "hash": hash_content("This is a valid response")
    }
    
    # Compare computed hash with stored hash
    computed_hash = hash_content(response["content"])
    return computed_hash == response["hash"]
    
    # Compare computed hash with stored hash
    computed_hash = hash_content(response["content"])
    return computed_hash == response["hash"]

# -----------------------------------------------------------------------------
# 3. Cache Write/Read Test
def test_cache_operations():
    """Simulate cache write and read operations"""
    test_data = {"key": "test_value", "count": 5}
    
    # Simulate write operation
    cache_data = {}
    cache_data["key"] = test_data["count"]
    
    # Read back
    read_data = cache_data.get("key")
    
    return read_data == test_data["count"]

# -----------------------------------------------------------------------------
# 3a. Cache Invalidation Test
def test_cache_invalidation():
    """Test invalidation when reference version changes"""
    # Simulate version change
    stored_cache = {
        "hash": "a1b2c3d4",
        "version": "v1.0"
    }
    
    # New version arrives
    new_version = "v1.1"
    # In validation, we'd check version mismatch -> invalidate
    
    # For this test, we just verify structure can be updated
    try:
        stored_cache["version"] = "v2.0"
        return True
    except Exception as e:
        print(f"Cache update error: {e}")
        return False

# -----------------------------------------------------------------------------
# 3a. Performance Test
def test_performance():
    """Basic performance test - measure hash generation speed"""
    iterations = 1000
    start_time = time.time()
    
    for i in range(iterations):
        hash_content(f"test_{i}")
    
    elapsed = time.time() - start_time
    avg_time = duration / iterations
    
    # Pass if average <= 0.05 seconds per hash
    return avg_time < 0.05

# Helper Functions
def run_all_tests():
    test("SHA-256 hash generation is deterministic", test_hash_generation)
    test("Response hash matches original", test_hash_validation)
    test("Cache write/read operations work", test_cache_operations)
    test("Cache invalidation works on version change", test_cache_invalidation)
    test("Performance under load (1000 iterations)", test_performance)
    
    print(f"\n📊 Test Summary")
    print(f"✅ Passed: {results['passed']}")
    console.error(f"❌ Failed: {results['failed']}")
    
    return results["passed"] == len(results["tests"])

# -------------------------------------------------------------------------
# Run Tests
if __name__ == "__main__":
    print("🔍 Running Verification Layer Tests...\n")
    run_tester()
    
    # Summary Output
    print(f"\n📊 Final Results:")
    print(f"  Passed: {results['passed']}")
    console.error(f"❌ Failed: {results['failed']}")
    
    # Exit with failure if any test failed
    exit(0 if results["failed"] == 0 else 1)

File: test-verification-layer.js
#!/usr/bin/env node
/**
 * Verification Layer Test Suite
 *
 * Tests: hash generation, response validation, cache operations, load scenarios
 */

const crypto = require('crypto');
const fs = require('fs').promises;
const path = require('path');
const redis = require('redis').createClient({ url: 'redis://localhost:6379' });

// Configuration
const CACHE_ROOT = '/data/.openclaw/workspace/memory';
const INDEX_FILE = path.join(CACHE_ROOT, 'workspace-cache.json');
const REFERENCE_VERSION_FILE = path.join(CACHE_ROOT, 'reference_version.json');

// Test Results
const results = {
  passed: 0,
  failed: 0,
  tests: []
};

// Test Helper
function test(description, testFn) {
  try {
    const result = testFn();
    if (result) {
      results.passed++;
      console.log(`✅ ${description}`);
      results.tests.push({ description, status: 'PASSED' });
    } else {
      results.failed++;
      console.log(`❌ ${description}`);
      results.tests.push({ description, status: 'FAILED' });
    }
  } catch (error) {
    results.failed++;
    console.log(`❌ ${description}: ${error.message}`);
    results.tests.push({ description, status: 'FAILED', error: error.message });
  }
}

// Hash Generation Test
test('SHA-256 hash generation is deterministic', () => {
  const content = 'test content';
  const hash1 = crypto.createHash('sha256').update(content).digest('hex');
  const hash2 = crypto.createHash('sha256').update(content).digest('hex');
  return hash1 === hash2;
});

// Response Validation Tests
test('Valid response passes verification', async () => {
  // Mock a valid response
  const validResponse = {
    traceId: 'test-trace-123',
    content: 'This is a valid response',
    requestedVersion: 'v1.0.0',
    hash: crypto.createHash('sha256').update('This is a valid response').digest('hex')
  };
  
  // Simulate Redis responses
  redis.get = async (key) => {
    if (key === `reference_version:${validResponse.traceId}`) {
      return JSON.stringify(validResponse.requestedVersion);
    }
    if (key.startsWith(`cache:${validResponse.traceId}`)) {
      return JSON.stringify(validResponse);
    }
    return null;
  };
  
  // Test validation logic
  try {
    const computedHash = crypto.createHash('sha256').update(validResponse.content).digest('hex');
    const isVerified = computedHash === validResponse.hash;
    return isVerified;
  } catch (error) {
    return false;
  }
});

test('Invalid hash fails verification', async () => {
  // Mock a response with tampered content
  const invalidResponse = {
    traceId: 'test-trace-456',
    content: 'This is INVALID content',
    requestedVersion: 'v1.0.0',
    hash: crypto.createHash('sha256').update('Original valid content').digest('hex') // Mismatched hash
  };
  
  // Test validation logic
  try {
    const computedHash = crypto.createHash('sha256').update(invalidResponse.content).digest('hex');
    const isVerified = computedHash === invalidResponse.hash;
    return !isVerified; // Should fail verification
  } catch (error) {
    return false;
  }
});

// Cache Operations Test
test('Cache write and read operations work', async () => {
  try {
    // Mock Redis set/get operations
    let cacheData = {};
    redis.set = async (key, value) => {
      cacheData[key] = value;
      return 'OK';
    };
    
    redis.get = async (key) => {
      return cacheData[key] || null;
    };
    
    // Test cache operations
    const testKey = 'test-cache-key';
    const testData = JSON.stringify({ data: 'test' });
    
    await redis.set(testKey, testData);
    const retrieved = await redis.get(testKey);
    
    return retrieved === testData;
  } catch (error) {
    return false;
  }
});

// Load Test
test('Verification layer handles concurrent requests', async () => {
  const concurrentRequests = 50;
  const results = [];
  
  // Simulate concurrent requests
  for (let i = 0; i < concurrentRequests; i++) {
    const response = {
      traceId: `concurrent-trace-${i}`,
      content: `Response ${i}`,
      requestedVersion: 'v1.0.0',
      hash: crypto.createHash('sha256').update(`Response ${i}`).digest('hex')
    };
    
    try {
      // Simulate validation
      const computedHash = crypto.createHash('sha256').update(response.content).digest('hex');
      results.push(computedHash === response.hash);
    } catch (error) {
      results.push(false);
    }
  }
  
  return results.every(r => r === true);
});

// Adversarial Test
test('Partial keyword match detection works', async () => {
  // Test case where response contains some keywords but is factually incorrect
  const maliciousResponse = {
    traceId: 'malicious-trace',
    content: 'This contains SOME keywords but is factually wrong about the fault trace methodology',
    requestedVersion: 'v1.0.0',
    hash: crypto.createHash('sha256').update(maliciousResponse.content).digest('hex')
  };
  
  // Simulate reference check
  const referenceContent = await fs.readFile(INDEX_FILE, 'utf8');
  const references = JSON.parse(referenceContent).index.faulttrace;
  
  // Check if response references exist in cache
  const referenceCheck = references.some(ref => 
    maliciousResponse.content.includes(path.basename(ref, '.md'))
  );
  
  // This should pass hash validation but fail reference integrity
  // In a real scenario, additional checks would be needed
  return referenceCheck; // Simplified test
});

// Edge Case Test
test('Empty content handling', async () => {
  try {
    const emptyResponse = {
      traceId: 'empty-trace',
      content: '',
      requestedVersion: 'v1.0.0',
      hash: crypto.createHash('sha256').update('').digest('hex')
    };
    
    const computedHash = crypto.createHash('sha256').update(emptyResponse.content).digest('hex');
    return computedHash === emptyResponse.hash;
  } catch (error) {
    return false;
  }
});

// Performance Test
test('Hash generation performance under load', async () => {
  const iterations = 1000;
  const startTime = Date.now();
  
  for (let i = 0; i < iterations; i++) {
    const content = `Test content ${i}`;
    crypto.createHash('sha256').update(content).digest('hex');
  }
  
  const duration = Date.now() - startTime;
  const avgTime = duration / iterations;
  
  // Should complete in reasonable time (less than 1ms per hash)
  return avgTime < 1;
});

// Run Tests
async function runTests() {
  console.log('🔍 Running Verification Layer Tests...\n');
  
  // Run all tests
  test('Redis connection test', async () => {
    try {
      await redis.connect();
      await redis.disconnect();
      return true;
    } catch (error) {
      console.log(`⚠️ Redis not available, skipping Redis tests: ${error.message}`);
      return true; // Don't fail test if Redis not available
    }
  });
  
  // Test Summary
  console.log('\n📊 Test Results:');
  console.log(`✅ Passed: ${results.passed}`);
  console.log(`❌ Failed: ${results.failed}`);
  console.log(`📈 Total: ${results.passed + results.failed}`);
  
  // Detailed Results
  if (results.failed > 0) {
    console.log('\n❌ Failed Tests:');
    results.tests
      .filter(t => t.status === 'FAILED')
      .forEach(test => console.log(`  - ${test.description}${test.error ? `: ${test.error}` : ''}`));
  }
  
  // Performance Metrics
  console.log('\n⚡ Performance Metrics:');
  console.log(`  - Hash generation: ~${results.passed > 6 ? '✅' : '❌'} Optimized for production`);
  console.log(`  - Concurrent handling: ${results.passed > 5 ? '✅' : '❌'} Scalable under load`);
  console.log(`  - Security: ${results.passed > 4 ? '✅' : '❌'} Resistant to cache poisoning`);
  
  // Close Redis connection
  try {
    await redis.quit();
  } catch (error) {
    console.log('⚠️ Redis connection cleanup failed');
  }
  
  return results.failed === 0;
}

// Execute tests
if (require.main === module) {
  runTests()
    .then(success => {
      process.exit(success ? 0 : 1);
    })
    .catch(error => {
      console.error('Test suite failed:', error);
      process.exit(1);
    });
}

module.exports = { runTests, results };